src/compiler/checker.ts

```diff
@@ -4211,8 +4211,12 @@ namespace ts {
                 context.flags &= ~NodeBuilderFlags.InTypeAlias;
 
                 if (!type) {
-                    context.encounteredError = true;
-                    return undefined!; // TODO: GH#18217
+                    if (!(context.flags & NodeBuilderFlags.AllowEmptyUnionOrIntersection)) {

```

Comment:
As we discussed offline, I think this (and below) would be clearer as `IgnoreErrors`, but I can live with this version if it's more idiomatic.

Reply:
Ah, I struggled to enumerate it earlier, but if we check for `NodeBuilderFlags.IgnoreErrors`, then if _any_ of the `AllowX` flags are set, the error would be suppressed (as that's how binary `&` works) - it's much neater to tie it to a specific flag (I'd have been tempted to add new ones, but we're actually entirely our of flag space in this enum).